diff -Naru freeciv-git-2026-02-24/Makefile.am freeciv-new-patched/Makefile.am
--- freeciv-git-2026-02-24/Makefile.am	2026-02-24 19:01:40.851552463 +0000
+++ freeciv-new-patched/Makefile.am	2026-02-24 19:09:00.747429296 +0000
@@ -25,7 +25,6 @@
 	tests		\
 	platforms	\
 	$(SSUBDIRS)	\
-	client		\
 	translations	\
 	doc		\
 	tools		\
diff -Naru freeciv-git-2026-02-24/common/featured_text.c freeciv-new-patched/common/featured_text.c
--- freeciv-git-2026-02-24/common/featured_text.c	2026-02-24 19:01:40.867552618 +0000
+++ freeciv-new-patched/common/featured_text.c	2026-02-24 19:07:10.977764817 +0000
@@ -34,8 +34,8 @@
 
 #include "featured_text.h"
 
-#define SEQ_START '['
-#define SEQ_STOP ']'
+#define SEQ_START '<'
+#define SEQ_STOP '>'
 #define SEQ_END '/'
 
 #define MAX_LEN_STR 32
@@ -74,7 +74,7 @@
 
 const struct ft_color ftc_warning       = FT_COLOR("#FF0000",   nullptr);
 const struct ft_color ftc_log           = FT_COLOR("#7F7F7F",   nullptr);
-const struct ft_color ftc_server        = FT_COLOR("#8B0000",   nullptr);
+const struct ft_color ftc_server        = FT_COLOR("#FFFFFF",   nullptr);
 const struct ft_color ftc_client        = FT_COLOR("#EF7F00",   nullptr);
 const struct ft_color ftc_editor        = FT_COLOR("#0000FF",   nullptr);
 const struct ft_color ftc_command       = FT_COLOR("#006400",   nullptr);
@@ -83,7 +83,7 @@
 const struct ft_color ftc_player_lost   = FT_COLOR("#FFFFFF",   "#000000");
 const struct ft_color ftc_game_start    = FT_COLOR("#00FF00",   "#115511");
 
-const struct ft_color ftc_chat_public   = FT_COLOR("#00008B",   nullptr);
+const struct ft_color ftc_chat_public   = FT_COLOR("#FFFFFF",   nullptr);
 const struct ft_color ftc_chat_ally     = FT_COLOR("#551166",   nullptr);
 const struct ft_color ftc_chat_private  = FT_COLOR("#A020F0",   nullptr);
 const struct ft_color ftc_chat_luaconsole = FT_COLOR("#006400", nullptr);
@@ -144,7 +144,7 @@
   case TTT_UNDERLINE:
     return "u";
   case TTT_COLOR:
-    return "c";
+    return "font";
   case TTT_LINK:
     return "l";
   };
@@ -243,7 +243,7 @@
     {
       if (!find_option(sequence, "foreground", ptag->color.foreground,
                        sizeof(ptag->color.foreground))
-          && !find_option(sequence, "fg", ptag->color.foreground,
+          && !find_option(sequence, "color", ptag->color.foreground,
                           sizeof(ptag->color.foreground))) {
         ptag->color.foreground[0] = '\0';
       }
@@ -487,7 +487,7 @@
                                text_tag_type_short_name(ptag->type));
 
       if (ptag->color.foreground[0] != '\0') {
-        ret += fc_snprintf(buf + ret, len - ret, " fg=\"%s\"",
+        ret += fc_snprintf(buf + ret, len - ret, " color=\"%s\"",
                            ptag->color.foreground);
       }
       if (ptag->color.background[0] != '\0') {
@@ -1076,10 +1076,9 @@
 {
   static char buf[MAX_LEN_LINK];
 
-  fc_snprintf(buf, sizeof(buf), "%c%s tgt=\"%s\" id=%d name=\"%s\" %c%c",
-              SEQ_START, text_tag_type_short_name(TTT_LINK),
-              text_link_type_name(TLT_CITY), pcity->id,
-              city_name_get(pcity), SEQ_END, SEQ_STOP);
+  fc_snprintf(buf, sizeof(buf), "<a href=\"#\" onclick=\"show_city_dialog_by_id(%d);\">%s</a>",
+              pcity->id,
+              city_name_get(pcity));
   return buf;
 }
 
diff -Naru freeciv-git-2026-02-24/common/map.h freeciv-new-patched/common/map.h
--- freeciv-git-2026-02-24/common/map.h	2026-02-24 19:01:40.869552637 +0000
+++ freeciv-new-patched/common/map.h	2026-02-24 19:09:14.384507891 +0000
@@ -655,7 +655,7 @@
 #ifdef FREECIV_WEB
 #define MAP_DEFAULT_SIZE         3
 #define MAP_MIN_SIZE             0
-#define MAP_MAX_SIZE             38
+#define MAP_MAX_SIZE             150
 #else  /* FREECIV_WEB */
 #define MAP_DEFAULT_SIZE         4
 #define MAP_MIN_SIZE             0
diff -Naru freeciv-git-2026-02-24/common/networking/packets.c freeciv-new-patched/common/networking/packets.c
--- freeciv-git-2026-02-24/common/networking/packets.c	2026-02-24 19:01:40.871552656 +0000
+++ freeciv-new-patched/common/networking/packets.c	2026-02-24 19:08:38.855303502 +0000
@@ -692,52 +692,6 @@
                                            packet_player_attribute_chunk
                                            *chunk)
 {
-  log_packet("received attribute chunk %u/%u %u",
-             (unsigned int) chunk->offset,
-             (unsigned int) chunk->total_length,
-             (unsigned int) chunk->chunk_length);
-
-  if (chunk->total_length < 0
-      || chunk->chunk_length < 0
-      || chunk->total_length >= MAX_ATTRIBUTE_BLOCK
-      || chunk->offset < 0
-      || chunk->offset > chunk->total_length /* necessary check on 32 bit systems */
-      || chunk->chunk_length > chunk->total_length
-      || chunk->offset + chunk->chunk_length > chunk->total_length
-      || (chunk->offset != 0
-          && chunk->total_length != pplayer->attribute_block_buffer.length)) {
-    /* wrong attribute data */
-    if (pplayer->attribute_block_buffer.data) {
-      free(pplayer->attribute_block_buffer.data);
-      pplayer->attribute_block_buffer.data = nullptr;
-    }
-    pplayer->attribute_block_buffer.length = 0;
-    log_error("Received wrong attribute chunk");
-    return;
-  }
-  /* first one in a row */
-  if (chunk->offset == 0) {
-    if (pplayer->attribute_block_buffer.data) {
-      free(pplayer->attribute_block_buffer.data);
-      pplayer->attribute_block_buffer.data = nullptr;
-    }
-    pplayer->attribute_block_buffer.data = fc_malloc(chunk->total_length);
-    pplayer->attribute_block_buffer.length = chunk->total_length;
-  }
-  memcpy((char *) (pplayer->attribute_block_buffer.data) + chunk->offset,
-         chunk->data, chunk->chunk_length);
-
-  if (chunk->offset + chunk->chunk_length == chunk->total_length) {
-    /* Received full attribute block */
-    if (pplayer->attribute_block.data != nullptr) {
-      free(pplayer->attribute_block.data);
-    }
-    pplayer->attribute_block.data = pplayer->attribute_block_buffer.data;
-    pplayer->attribute_block.length = pplayer->attribute_block_buffer.length;
-
-    pplayer->attribute_block_buffer.data = nullptr;
-    pplayer->attribute_block_buffer.length = 0;
-  }
 }
 
 /**********************************************************************//**
diff -Naru freeciv-git-2026-02-24/common/networking/packets.def freeciv-new-patched/common/networking/packets.def
--- freeciv-git-2026-02-24/common/networking/packets.def	2026-02-24 19:01:40.871552656 +0000
+++ freeciv-new-patched/common/networking/packets.def	2026-02-24 19:06:18.775437978 +0000
@@ -486,6 +486,7 @@
   SINT16 altitude;
   STRING spec_sprite[MAX_LEN_NAME];
   STRING label[MAX_LEN_MAP_LABEL];
+  UINT32 height;
 end
 
 # The variables in the packet are listed in alphabetical order.
@@ -2475,3 +2476,27 @@
 
   BV_ACTIONS utype_actions;
 end
+
+# Used for showing GOTO path in web client.
+PACKET_WEB_GOTO_PATH_REQ = 287; cs, dsend, handle-via-fields
+  UNIT unit_id;
+  TILE goal;
+end
+
+PACKET_WEB_GOTO_PATH = 288; sc, dsend, handle-via-fields, no-handle
+  UNIT unit_id;
+  UINT16 length;
+  DIRECTION dir[MAX_LEN_ROUTE:length];
+  TILE dest;
+  UINT32 turns;
+end
+
+PACKET_WEB_INFO_TEXT_REQ = 289; cs, dsend, handle-via-fields
+  TILE loc;
+  UNIT visible_unit;
+  UNIT focus_unit;
+end
+
+PACKET_WEB_INFO_TEXT_MESSAGE = 290; sc, dsend, handle-via-fields, no-handle
+  STRING message[MAX_LEN_MSG];
+end
diff -Naru freeciv-git-2026-02-24/common/networking/packets_json.h freeciv-new-patched/common/networking/packets_json.h
--- freeciv-git-2026-02-24/common/networking/packets_json.h	2026-02-24 19:01:40.871552656 +0000
+++ freeciv-new-patched/common/networking/packets_json.h	2026-02-24 19:06:26.715483163 +0000
@@ -53,8 +53,6 @@
       }                                                                 \
       size = dio_output_used(&dout.raw);                                \
                                                                         \
-      dio_output_rewind(&(dout.raw));                                   \
-      dio_put_uint16_raw(&(dout.raw), size);                            \
       free(json_buffer);                                                \
       json_decref(dout.json);                                           \
     } else {                                                            \
diff -Naru freeciv-git-2026-02-24/common/tile.h freeciv-new-patched/common/tile.h
--- freeciv-git-2026-02-24/common/tile.h	2026-02-24 19:01:40.877552714 +0000
+++ freeciv-new-patched/common/tile.h	2026-02-24 19:06:32.617520983 +0000
@@ -65,6 +65,7 @@
   int altitude;
   char *label;                          /* nullptr for no label */
   char *spec_sprite;
+  int height;
 };
 
 /* 'struct tile_list' and related functions. */
diff -Naru freeciv-git-2026-02-24/fc_version freeciv-new-patched/fc_version
--- freeciv-git-2026-02-24/fc_version	2026-02-24 19:01:41.947563078 +0000
+++ freeciv-new-patched/fc_version	2026-02-24 19:05:57.312301617 +0000
@@ -76,7 +76,7 @@
 # On FREECIV_DEBUG builds, optional capability "debug" gets automatically
 # appended to this.
 #
-NETWORK_CAPSTRING="+Freeciv.Devel-${MAIN_VERSION}-2026.Feb.05"
+NETWORK_CAPSTRING="+Freeciv.Web.Devel-3.3"
 
 # If you are distributing freeciv, and apply any patches at all,
 # patch also this field to contain your identification.
diff -Naru freeciv-git-2026-02-24/server/auth.c freeciv-new-patched/server/auth.c
--- freeciv-git-2026-02-24/server/auth.c	2026-02-24 19:01:41.957563175 +0000
+++ freeciv-new-patched/server/auth.c	2026-02-24 19:07:30.175886316 +0000
@@ -74,6 +74,18 @@
 {
   char tmpname[MAX_LEN_NAME] = "\0";
 
+#ifdef FREECIV_WEB
+  if (srvarg.server_password_enabled) {
+    char buffer[MAX_LEN_MSG];
+
+    fc_snprintf(buffer, sizeof(buffer), _("Enter password for this private game:"));
+    dsend_packet_authentication_req(pconn, AUTH_LOGIN_FIRST, buffer);
+    pconn->server.auth_settime = time(NULL);
+
+    return TRUE;
+  }
+#endif /* FREECIV_WEB */
+
   /* Assign the client a unique guest name/Reject if guests aren't allowed */
   if (is_guest_name(username)) {
     if (srvarg.auth_allow_guests) {
@@ -157,6 +169,32 @@
 {
   char msg[MAX_LEN_MSG];
 
+#ifdef FREECIV_WEB
+  if (srvarg.server_password_enabled) {
+    if (pconn->server.status == AS_ESTABLISHED) {
+      notify_conn(NULL, NULL, E_SETTING, ftc_server,
+                  _("Server password already set."));
+    } else if (!strncmp(srvarg.server_password, password, MAX_LEN_PASSWORD)) {
+      establish_new_connection(pconn);
+    } else {
+      char buffer[MAX_LEN_MSG];
+
+      fc_snprintf(buffer, sizeof(buffer),
+                  _("Incorrect password entered. Please try again."));
+      dsend_packet_authentication_req(pconn, AUTH_LOGIN_FIRST, buffer);
+      pconn->server.auth_settime = time(NULL);
+    }
+
+    return TRUE;
+  } else {
+    srvarg.server_password_enabled = TRUE;
+    sz_strlcpy(srvarg.server_password, password);
+    notify_conn(NULL, NULL, E_SETTING, ftc_server, _("Server password set."));
+
+    return TRUE;
+  }
+#endif /* FREECIV_WEB */
+
   if (pconn->server.status == AS_REQUESTING_NEW_PASS) {
 
     /* check if the new password is acceptable */
diff -Naru freeciv-git-2026-02-24/server/commands.c freeciv-new-patched/server/commands.c
--- freeciv-git-2026-02-24/server/commands.c	2026-02-24 19:01:41.959563194 +0000
+++ freeciv-new-patched/server/commands.c	2026-02-24 19:09:28.949595796 +0000
@@ -543,7 +543,7 @@
       "all cities and units etc. Use with care!"), NULL,
    CMD_ECHO_ALL, VCF_NONE, 50
   },
-  {"save",	ALLOW_ADMIN,
+  {"save",	ALLOW_CTRL,
    /* TRANS: translate text between <> only */
    N_("save\n"
       "save <file-name>"),
diff -Naru freeciv-git-2026-02-24/server/meta.c freeciv-new-patched/server/meta.c
--- freeciv-git-2026-02-24/server/meta.c	2026-02-24 19:01:41.962563224 +0000
+++ freeciv-new-patched/server/meta.c	2026-02-24 19:07:52.230025063 +0000
@@ -324,7 +324,8 @@
                          get_meta_message_string());
 
     /* NOTE: Send info for ALL players or none at all. */
-    if (normal_player_count() == 0) {
+    // FIXME: there is a bug here, if there are players then the server isn't reported to the metaserver.
+    if (TRUE || normal_player_count() == 0) {
       netfile_add_form_int(post, "dropplrs", 1);
     } else {
       players = 0; /* Counter for players_available */
@@ -391,6 +392,10 @@
         }
       } players_iterate_end;
 
+      if (S_S_INITIAL == server_state()) {
+        humans = conn_list_size(game.est_connections);
+      }
+
       /* Send the number of available players. */
       netfile_add_form_int(post, "available", players);
       netfile_add_form_int(post, "humans", humans);
@@ -533,7 +538,7 @@
     since_previous = timer_read_seconds(last_send_timer);
 
     /* Don't allow the user to spam the metaserver with updates */
-    if (since_previous < METASERVER_MIN_UPDATE_INTERVAL) {
+    if (flag != META_FORCE && since_previous < METASERVER_MIN_UPDATE_INTERVAL) {
       if (flag == META_INFO) {
         want_update = TRUE; /* We couldn't update now, but update a.s.a.p. */
       }
diff -Naru freeciv-git-2026-02-24/server/meta.h freeciv-new-patched/server/meta.h
--- freeciv-git-2026-02-24/server/meta.h	2026-02-24 19:01:41.962563224 +0000
+++ freeciv-new-patched/server/meta.h	2026-02-24 19:07:59.827072017 +0000
@@ -19,13 +19,14 @@
 
 #define DEFAULT_META_SERVER_NO_SEND  TRUE
 #define DEFAULT_META_SERVER_ADDR     FREECIV_META_URL
-#define METASERVER_REFRESH_INTERVAL   (3*60)
+#define METASERVER_REFRESH_INTERVAL   60
 #define METASERVER_MIN_UPDATE_INTERVAL 7   /* not too short, not too long */
 
 enum meta_flag {
   META_INFO,
   META_REFRESH,
-  META_GOODBYE
+  META_GOODBYE,
+  META_FORCE
 };
 
 const char *default_meta_patches_string(void);
diff -Naru freeciv-git-2026-02-24/server/stdinhand.c freeciv-new-patched/server/stdinhand.c
--- freeciv-git-2026-02-24/server/stdinhand.c	2026-02-24 19:01:41.972563320 +0000
+++ freeciv-new-patched/server/stdinhand.c	2026-02-24 19:08:13.368155555 +0000
@@ -639,11 +639,6 @@
 **************************************************************************/
 static bool save_command(struct connection *caller, char *arg, bool check)
 {
-  if (is_restricted(caller)) {
-    cmd_reply(CMD_SAVE, caller, C_FAIL,
-              _("You cannot save games manually on this server."));
-    return FALSE;
-  }
   if (!check) {
     save_game(arg, "User request", FALSE);
   }
